### PR TITLE
Architecture as Code: Microservices blog post

### PR DESCRIPTION
**Pull request by spara**

### CODE REVIEW

File: content/blog/architecture-as-code-microservices/index.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    enableFlowLogs?: boolean;
+
+    endpoints: {
+        s3: boolean;
+        dynamodb: boolean;
+    };
+}
+```
+
+Finally, we need to be able to peer VPCs so that the services can communicate with each other. We can take advantage of Typescript's [member function](https://www.typescriptlang.org/docs/handbook/classes.html#public-private-and-protected-modifiers) to implement peering.
```
Comment (justinvp) [inline content/blog/architecture-as-code-microservices/index.md:75]: Nit:

```suggestion
Finally, we need to be able to peer VPCs so that the services can communicate with each other. We can take advantage of TypeScript's [member function](https://www.typescriptlang.org/docs/handbook/classes.html#public-private-and-protected-modifiers) to implement peering.
```

File: content/blog/architecture-as-code-microservices/index.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+Monolithic stacks are a common way to structure a project. A single project defines the infrastructure and resources for an entire service. However, Pulumi also supports deploying microservices as individual stacks within a project. Because microservices represent a specific business capability, having stacks for each service allows teams to work independently of each other. Separate stacks can enforce security policies through Role-Based Access Control, and each service can be built and deployed independently of other components.
+
+Microservices deployed as stacks communicate with each other through the StackReference resource, which makes stack exports such as networking, environmental variables, and even Kubernetes configuration available to stacks in the project. The `StackReference` constructor takes as input a string of the form `<organization>/<project>/<stack>`, and lets services access the outputs of that stack. The following section walks through a microservices example that deploys a database on AWS RDS, a REST application in AWS Fargate, and an AWS Application Load Balancer.
```
Comment (metral) [inline content/blog/architecture-as-code-microservices/index.md:30]:
```suggestion
Microservices deployed as stacks communicate with each other through the [StackReference](https://www.pulumi.com/docs/intro/concepts/organizing-stacks-projects/#inter-stack-dependencies) resource, which makes stack exports such as networking, environmental variables, and even Kubernetes configuration available to stacks in the project. The `StackReference` constructor takes as input a string of the form `<organization>/<project>/<stack>`, and lets services access the outputs of that stack. The following section walks through a microservices example that deploys a database on AWS RDS, a REST application in AWS Fargate, and an AWS Application Load Balancer.
```

File: content/blog/architecture-as-code-microservices/index.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+We'll use the [AWS Stack Reference example](https://github.com/pulumi/examples/tree/master/aws-stackreference-architecture) to illustrate how Pulumi uses Architecture as Code to create reusable components and deployment frameworks. We’ve previously [blogged]({{< relref "blog/architect-aws-application-infra-with-pulumi-stack-references" >}}) about this example and covered the code in some detail.
+
+Let's examine how it creates reusable components for building out resources. First up is the [VPC class](https://github.com/pulumi/examples/blob/master/aws-stackreference-architecture/networking/src/vpc.ts) in the [networking stack](https://github.com/pulumi/examples/tree/master/aws-stackreference-architecture/networking).
+
+In the networking service, [`vpc.ts](https://github.com/pulumi/examples/blob/master/aws-stackreference-architecture/networking/src/vpc.ts) creates a TypeScript class that extends Pulumi ComponentResource. A ComponentResource abstracts one or more children that do not require custom create, read, update, and delete operations for provisioning. We can add the related resources to the correct parent to build out the class.
```
Comment (metral) [inline content/blog/architecture-as-code-microservices/index.md:38]:
```suggestion
In the networking service, [vpc.ts](https://github.com/pulumi/examples/blob/master/aws-stackreference-architecture/networking/src/vpc.ts) creates a TypeScript class that extends Pulumi [ComponentResource](https://www.pulumi.com/docs/intro/concepts/programming-model/#resources). A ComponentResource abstracts one or more children that do not require custom create, read, update, and delete operations for provisioning. We can add the related resources to the correct parent to build out the class.
```

File: content/blog/architecture-as-code-microservices/index.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+```
+
+Finally, we need to be able to peer VPCs so that the services can communicate with each other. We can take advantage of TypeScript's [member function](https://www.typescriptlang.org/docs/handbook/classes.html#public-private-and-protected-modifiers) to implement peering.
+
+```ts
+public configurePeering(args: PeerToArgs) {
+
+  // …
+}
+```
```
Comment (metral) [inline content/blog/architecture-as-code-microservices/index.md:82]: The code blocks here don't seem related to ComponentResources, do they?
It appears that we can clean this up

File: content/blog/architecture-as-code-microservices/index.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+$ cd microservices
+
+$ git init
+$ git remote add origin -f https://github.com/pulumi/examples/
+$ git config core.sparseCheckout true
+$ aws-stackreference-architecture >> .git/info/sparse-checkout
+$ git pull origin master
+```
+
+Follow the instructions to deploy the application. One caveat, when deploying the network microservice, set the number of availability zones to a minimum of two, e.g., `pulumi config set azCount 2`. Experiment with the example by adding data to the database and adding your REST application as a container. Add a web client to submit requests and display them. Let us know if you have questions on [Community Slack](https://pulumi-community.slack.com/) or send a tweet to show off your application.
```
Comment (metral) [inline content/blog/architecture-as-code-microservices/index.md:101]:
```suggestion
Follow the instructions to deploy the application. Experiment with the example by adding data to the database and adding your REST application as a container. Add a web client to submit requests and display them. Let us know if you have questions on [Community Slack](https://pulumi-community.slack.com/) or send a tweet to show off your application.
```


The `azCount` defaults to 2 in the networking stack:
https://github.com/pulumi/examples/blob/90f6fa541f7d6cf0b5171b39e7ef234b785cc98f/aws-stackreference-architecture/networking/src/index.ts#L7

**Comment by metral** [comment]
Great post! Left a couple of nits but LGTM 👍

File: content/blog/architecture-as-code-microservices/index.md (hunk 2 above)
Comment (justinvp) [reply content/blog/architecture-as-code-microservices/index.md:30]: Let's use relref links for these.

File: content/blog/architecture-as-code-microservices/index.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+Microservices deployed as stacks communicate with each other through the [StackReference](https://www.pulumi.com/docs/intro/concepts/organizing-stacks-projects/#inter-stack-dependencies) resource, which makes stack exports such as networking, environmental variables, and even Kubernetes configuration available to stacks in the project. The `StackReference` constructor takes as input a string of the form `<organization>/<project>/<stack>`, and lets services access the outputs of that stack. The following section walks through a microservices example that deploys a database on AWS RDS, a REST application in AWS Fargate, and an AWS Application Load Balancer.
```
Comment (justinvp) [inline content/blog/architecture-as-code-microservices/index.md:30]: relref

File: content/blog/architecture-as-code-microservices/index.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+Let's examine how it creates reusable components for building out resources. First up is the [VPC class](https://github.com/pulumi/examples/blob/master/aws-stackreference-architecture/networking/src/vpc.ts) in the [networking stack](https://github.com/pulumi/examples/tree/master/aws-stackreference-architecture/networking).
+
+In the networking service, [vpc.ts](https://github.com/pulumi/examples/blob/master/aws-stackreference-architecture/networking/src/vpc.ts) creates a TypeScript class that extends Pulumi [ComponentResource](https://www.pulumi.com/docs/intro/concepts/programming-model/#resources). A ComponentResource abstracts one or more children that do not require custom create, read, update, and delete operations for provisioning. We can add the related resources to the correct parent to build out the class.
```
Comment (justinvp) [inline content/blog/architecture-as-code-microservices/index.md:38]: relref

File: content/blog/architecture-as-code-microservices/index.md (hunk 6 above)
Comment (isaac-pulumi) [inline content/blog/architecture-as-code-microservices/index.md:30]:
```suggestion
Microservices deployed as stacks communicate with each other through the [StackReference]({{<relref "/docs/intro/concepts/organizing-stacks-projects/#inter-stack-dependencies" >}}) resource, which makes stack exports such as networking, environmental variables, and even Kubernetes configuration available to stacks in the project. The `StackReference` constructor takes as input a string of the form `<organization>/<project>/<stack>`, and lets services access the outputs of that stack. The following section walks through a microservices example that deploys a database on AWS RDS, a REST application in AWS Fargate, and an AWS Application Load Balancer.
```

File: content/blog/architecture-as-code-microservices/index.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+date: 2020-04-07
+meta_desc: "Build microservices with reusable components created with common programming languages and Pulumi"
+meta_image: microservices.png
+authors:
+    - sophia-parafina
+tags:
+    - microservices
+---
+
+This article is the third in a series about Architecture as Code. The [first article]({{ relref "/blog/architecture-as-code-intro">}}) provided an overview of virtual machines, microservices, serverless, and Kubernetes. The [second]({{< relref "/blog/architecture-as-code-vm" >}}) one went in-depth on deploying virtual machines as reusable components. In this third installment, we'll look at microservices and how to implement them as reusable components with Pulumi.
```
Comment (isaac-pulumi) [inline content/blog/architecture-as-code-microservices/index.md:12]:
```suggestion
This article is the third in a series about Architecture as Code. The [first article]({{< relref "/blog/architecture-as-code-intro">}}) provided an overview of virtual machines, microservices, serverless, and Kubernetes. The [second]({{< relref "/blog/architecture-as-code-vm" >}}) one went in-depth on deploying virtual machines as reusable components. In this third installment, we'll look at microservices and how to implement them as reusable components with Pulumi.
```

File: content/blog/architecture-as-code-microservices/index.md (hunk 6 above)
Comment (isaac-pulumi) [inline content/blog/architecture-as-code-microservices/index.md:30]:
```suggestion
Microservices deployed as stacks communicate with each other through the [StackReference]({{< relref "/docs/intro/concepts/organizing-stacks-projects/#inter-stack-dependencies" >}}) resource, which makes stack exports such as networking, environmental variables, and even Kubernetes configuration available to stacks in the project. The `StackReference` constructor takes as input a string of the form `<organization>/<project>/<stack>`, and lets services access the outputs of that stack. The following section walks through a microservices example that deploys a database on AWS RDS, a REST application in AWS Fargate, and an AWS Application Load Balancer.
```

File: content/blog/architecture-as-code-microservices/index.md (hunk 7 above)
Comment (isaac-pulumi) [inline content/blog/architecture-as-code-microservices/index.md:38]:
```suggestion
In the networking service, [vpc.ts](https://github.com/pulumi/examples/blob/master/aws-stackreference-architecture/networking/src/vpc.ts) creates a TypeScript class that extends Pulumi [ComponentResource]({{< relref "/docs/intro/concepts/programming-model/#resources" >}}). A ComponentResource abstracts one or more children that do not require custom create, read, update, and delete operations for provisioning. We can add the related resources to the correct parent to build out the class.
```

File: content/blog/architecture-as-code-microservices/index.md (hunk 8 above)
Comment (justinvp) [reply content/blog/architecture-as-code-microservices/index.md:12]: Let's fix the space at the end too:

```suggestion
This article is the third in a series about Architecture as Code. The [first article]({{< relref "/blog/architecture-as-code-intro" >}}) provided an overview of virtual machines, microservices, serverless, and Kubernetes. The [second]({{< relref "/blog/architecture-as-code-vm" >}}) one went in-depth on deploying virtual machines as reusable components. In this third installment, we'll look at microservices and how to implement them as reusable components with Pulumi.
```

File: content/blog/architecture-as-code-microservices/index.md (hunk 6 above)
Comment (justinvp) [reply content/blog/architecture-as-code-microservices/index.md:30]: The way the anchor link works with relref needs to be fixed:

```suggestion
Microservices deployed as stacks communicate with each other through the [StackReference]({{< relref "/docs/intro/concepts/organizing-stacks-projects#inter-stack-dependencies" >}}) resource, which makes stack exports such as networking, environmental variables, and even Kubernetes configuration available to stacks in the project. The `StackReference` constructor takes as input a string of the form `<organization>/<project>/<stack>`, and lets services access the outputs of that stack. The following section walks through a microservices example that deploys a database on AWS RDS, a REST application in AWS Fargate, and an AWS Application Load Balancer.
```

File: content/blog/architecture-as-code-microservices/index.md (hunk 7 above)
Comment (justinvp) [reply content/blog/architecture-as-code-microservices/index.md:38]: Similarly:

```suggestion
In the networking service, [vpc.ts](https://github.com/pulumi/examples/blob/master/aws-stackreference-architecture/networking/src/vpc.ts) creates a TypeScript class that extends Pulumi [ComponentResource]({{< relref "/docs/intro/concepts/programming-model#resources" >}}). A ComponentResource abstracts one or more children that do not require custom create, read, update, and delete operations for provisioning. We can add the related resources to the correct parent to build out the class.
```
